src/libs/safe-path/src/safe_path_buf.rs

```diff
@@ -0,0 +1,125 @@
+    target: PathBuf,
+}
+
+impl SafePathBuf {
+    /// Create a `SafePathBuf` from the `root` and an unsafe `path`.
+    ///
+    /// The `path` must be a subdirectory of `root`, otherwise error will be returned.
+    pub fn new<R: AsRef<Path>, U: AsRef<Path>>(root: R, path: U) -> Result<Self> {
+        let safe_path = safe_join(root, path)?;
+        Self::from_path(safe_path)
+    }
+
+    /// Create a `SafePathBuf` from an path.
+    ///
+    /// If the resolved value of `path` doesn't equal to `path`, an error will be returned.
+    pub fn from_path<P: AsRef<Path>>(path: P) -> Result<Self> {
+        let file = open_by_path(path.as_ref())?;
+        let proc_path = format!("/proc/self/fd/{}", file.as_raw_fd());
+        let link_path = fs::read_link(&proc_path)?;
+
+        if link_path.as_path() != path.as_ref() {
+            Err(Error::new(
+                ErrorKind::Other,
+                format!(
+                    "The target path changes from {} to {} underneath, possible under attacking!!!",
+                    path.as_ref().display(),
+                    link_path.display()
+                ),
+            ))
+        } else {
+            Ok(SafePathBuf {
+                file,
+                path: PathBuf::from(proc_path),
+                target: link_path,
+            })
+        }
+    }
+
+    /// Get the real target path.
+    pub fn target(&self) -> &Path {
+        &self.target
+    }
+
+    /// Check whether the target path is a directory.
+    pub fn is_dir(&self) -> bool {
+        self.target.is_dir()
+    }
+}
+
+impl Deref for SafePathBuf {
+    type Target = PathBuf;
+
+    fn deref(&self) -> &Self::Target {
+        &self.path
+    }
+}
+
+impl AsRef<Path> for SafePathBuf {
+    fn as_ref(&self) -> &Path {
+        self.path.as_path()
+    }
+}
+
+impl AsRawFd for SafePathBuf {
+    fn as_raw_fd(&self) -> RawFd {
+        self.file.as_raw_fd()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_safe_path_buf() {
+        let rootfs_dir = tempfile::tempdir().expect("failed to create tmpdir");
+        let rootfs_path = rootfs_dir.path();
+
+        std::fs::create_dir(rootfs_path.join("symlink_dir")).unwrap();
+        std::os::unix::fs::symlink("/endpoint", rootfs_path.join("symlink_dir/endpoint")).unwrap();
+        fs::write(rootfs_path.join("endpoint"), "test").unwrap();
+        let path = SafePathBuf::new(rootfs_path, "symlink_dir/endpoint").unwrap();
+        let link = fs::read_link(&path).unwrap();
+        assert_eq!(link, rootfs_path.join("endpoint"));
+        let content = fs::read_to_string(&path).unwrap();
+        assert_eq!(&content, "test");
+
+        let path = SafePathBuf::from_path(rootfs_path.join("endpoint")).unwrap();
+        let link = fs::read_link(path.deref()).unwrap();
+        assert_eq!(link, rootfs_path.join("endpoint"));
+        let content = fs::read_to_string(&path).unwrap();
+        assert_eq!(&content, "test");
+    }
+}
```

Comment:
You never actually test the cases you're trying to protect against where a symlink is modified between the construction and use of the `SafePathBuf`.

Reply:
add test case for it.